provider/oci/storage_volumes.go

```diff
@@ -4,44 +4,568 @@
+	}
+	results := make([]storage.CreateVolumesResult, len(params))
+	instanceMap := map[instance.Id]*ociInstance{}
+	for i, volume := range params {
+		vol, err := v.createVolume(volume, instanceMap)
+		if err != nil {
+			results[i].Error = errors.Trace(err)
+			continue
+		}
+		results[i].Volume = vol
+	}
+	return results, nil
 }
 
-func (v *volumeSource) ReleaseVolumes(ctx context.ProviderCallContext, volIds []string) ([]error, error) {
-	return nil, errors.NotImplementedf("ReleaseVolumes")
+func (v *volumeSource) allVolumes() (map[string]ociCore.Volume, error) {
+	result := map[string]ociCore.Volume{}
+	request := ociCore.ListVolumesRequest{
+		CompartmentId: v.env.ecfg().compartmentID(),
+	}
+	response, err := v.storageAPI.ListVolumes(context.Background(), request)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, val := range response.Items {
+		if t, ok := val.FreeformTags[tags.JujuModel]; !ok {
+			continue
+		} else {
+			if t != "" && t != v.modelUUID {
+				continue
+			}
+		}
+		result[*val.Id] = val
+	}
+	return result, nil
+}
+
+func (v *volumeSource) ListVolumes(ctx envcontext.ProviderCallContext) ([]string, error) {
+	ids := []string{}
+	volumes, err := v.allVolumes()
+	if err != nil {
+		return nil, err
+	}
+
+	for k := range volumes {
+		ids = append(ids, k)
+	}
+	return ids, nil
+}
+
+func (v *volumeSource) DescribeVolumes(ctx envcontext.ProviderCallContext, volIds []string) ([]storage.DescribeVolumesResult, error) {
+	result := make([]storage.DescribeVolumesResult, len(volIds), len(volIds))
+
+	allVolumes, err := v.allVolumes()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	for i, val := range volIds {
+		if volume, ok := allVolumes[val]; ok {
+			volumeInfo := makeVolumeInfo(volume)
+			result[i].VolumeInfo = &volumeInfo
+		} else {
+			result[i].Error = errors.NotFoundf("%s", volume)
+		}
+	}
+	return result, nil
+}
+
+func (v *volumeSource) DestroyVolumes(ctx envcontext.ProviderCallContext, volIds []string) ([]error, error) {
+	volumes, err := v.allVolumes()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	errs := make([]error, len(volIds))
+
+	for idx, volId := range volIds {
+		volumeDetails, ok := volumes[volId]
+		if !ok {
+			errs[idx] = errors.NotFoundf("no such volume %s", volId)
+			continue
+		}
+		request := ociCore.DeleteVolumeRequest{
+			VolumeId: volumeDetails.Id,
+		}
+
+		response, err := v.storageAPI.DeleteVolume(context.Background(), request)
+		if err != nil && !v.env.isNotFound(response.RawResponse) {
+			errs[idx] = errors.Trace(err)
+			continue
+		}
+		err = v.env.waitForResourceStatus(
+			v.getVolumeStatus, volumeDetails.Id,
+			string(ociCore.VolumeLifecycleStateTerminated),
+			5*time.Minute)
+		if err != nil && !errors.IsNotFound(err) {
+			errs[idx] = errors.Trace(err)
+		} else {
+			errs[idx] = nil
+		}
+	}
+	return errs, nil
+}
+
+func (v *volumeSource) ReleaseVolumes(ctx envcontext.ProviderCallContext, volIds []string) ([]error, error) {
```

Comment:
I know these are interface implementations, with comments on the interface itself, but it would be good to get comments on all these public methods - something to summarise the actions taken. In this case that releasing is disassociating the volumes from the model and controller by removing them from the volume tag collections.